### PR TITLE
Fix atomic IME text replacement

### PR DESCRIPTION
**Pull request by RasulOs**
## Summary

- replace the complete field atomically with one selection plus `commitText` for `clear=true`
- reject partial, nonzero-offset, or end-unverified extracted windows before full-field replacement
- finish active composition before reading the final pre-commit snapshot, then select and replace the field
- verify append operations against the exact result predicted from the original text and selection
- never recommit or accessibility-replay any operation after `commitText` accepts it
- distinguish thrown `commitText` outcomes from observed acceptance and skip replay when delivery is unknown
- contain readback exceptions inside the editor and report accepted-but-unverifiable only after observed acceptance
- preserve one-shot appends when extracted text or composition preparation is unsupported
- preserve accepted and input-session-changed clear outcomes so `keyboardClear` cannot replay them through accessibility
- keep verification, pre-commit retries, and fallback bound to the original input session
- advance the session generation for every `onStartInput`, including editor restarts
- ignore app-added U+200B/U+FEFF sentinels for ordinary input, while requiring explicitly requested zero-width characters to survive
- route `clearText()` through the same verified replacement path

## Reproduction matrix

| Path | Before | After |
| --- | --- | --- |
| Accepted append with stale or throwing readback | The append could be committed twice or replayed through accessibility | Returns accepted-but-unverifiable after exactly one commit |
| Thrown `commitText` | An exception was treated as accepted even though delivery was unknown | Returns commit-outcome-unknown; the API reports an error and skips replay |
| Accepted clear with stale readback | A retry could select the old length and corrupt a longer replacement | Never recommits; returns accepted-but-unverifiable unless the first readback verifies |
| Accepted clear with unreadable readback | Boolean collapse could trigger accessibility on a newly focused field | Preserves accepted-but-unverifiable and skips fallback |
| Active composing region | `commitText` could replace only the composing range | Finishes composition before reading, selecting, and replacing the complete field |
| Unsupported append preparation | Text could be dropped even though `commitText` was supported | Performs one direct commit at the current selection and skips fallback when accepted |
| Snapshot-less append | Input was rejected before `commitText` | Commits once at the editor's current selection and skips fallback |
| Focus changes during input | Retry or fallback could modify a newly focused field | Returns an input-session-changed error without retry or fallback |
| Editor `restartInput` | A replacement connection retained the previous generation | Every start callback invalidates in-flight work, including restarts |
| Partial or truncated extracted text | `clear=true` could replace only the extracted slice | Rejects before committing unless the snapshot starts at zero, is a full report, and its represented suffix reaches the field end |
| Requested zero-width text | A dropped requested character could be normalized into false success | Requested U+200B/U+FEFF must be present exactly to verify |
| Android 16 Settings search | Legacy input could report success without atomic verification | Shared patched editor produced exactly `MRTEST7319`; the dedicated clear endpoint then emptied the field |

No search result was selected. Test text was cleared afterward, and the public Mobilerun IME/accessibility configuration was restored.

## Tests

- PASS: focused `ApiHandlerTest` and `InputConnectionTextEditorTest`
- PASS: `./gradlew assembleDebug --no-daemon`
- Full suite: 480/481 passed. The only failure is the unchanged pre-existing `SwitchTintResourceTest.trackTintSeparatesCheckedAndUncheckedStates` assertion.
- PASS on Android 16 `emulator-5554` through the mirrored internal debug package at the latest commit: composition-aware atomic replacement and the dedicated clear endpoint were verified from fresh UI hierarchies.
- The public debug APK was not installed over the existing public build because Android rejected its lower version code.

## Companion PR

- [Internal Portal #72](https://github.com/droidrun/mobilerun-portal-internal/pull/72)

### CODE REVIEW

File: app/src/main/java/com/mobilerun/portal/input/InputConnectionTextEditor.kt [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+                return TextInputResult.AcceptedUnverified
+            }
+
+            if (matches(after.text, text, clear)) {
+                return TextInputResult.Verified
+            }
+
+            val restarted = generationProvider() != generation || afterConnection !== connection
+            if (attempt < MAX_ATTEMPTS - 1 && (clear || restarted || !contains(after.text, text))) {
+                return@repeat
```
[the remark this answers is not in the document]
Comment (RasulOs) [reply app/src/main/java/com/mobilerun/portal/input/InputConnectionTextEditor.kt]: Fixed in c8d1d3d. Once an append commit is accepted, the editor now either verifies the exact expected post-state or returns AcceptedUnverified; it never retries or invokes accessibility fallback. Added stale-readback and unreadable-post-state regression tests that assert exactly one commit.

File: app/src/main/java/com/mobilerun/portal/input/InputConnectionTextEditor.kt [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+            startOffset = extracted.startOffset,
+            selectionStart = relativeStart,
+            selectionEnd = relativeEnd,
+        )
+    }
+
+    private fun selectionFor(snapshot: Snapshot, clear: Boolean): Pair<Int, Int> {
+        val relativeStart = if (clear) 0 else minOf(snapshot.selectionStart, snapshot.selectionEnd)
+        val relativeEnd = if (clear) snapshot.text.length else maxOf(snapshot.selectionStart, snapshot.selectionEnd)
+        return Pair(snapshot.startOffset + relativeStart, snapshot.startOffset + relativeEnd)
```
[the remark this answers is not in the document]
Comment (RasulOs) [reply app/src/main/java/com/mobilerun/portal/input/InputConnectionTextEditor.kt:260]: Fixed in c8d1d3d. Full-field replacement now requires startOffset == 0 and partialStartOffset == -1 before committing. Partial or offset windows are retried once before any commit, then rejected for safe accessibility fallback. Both conditions have regression tests.

File: app/src/main/java/com/mobilerun/portal/input/MobilerunKeyboardIME.kt [hunk 3]
```diff
@@ -126,10 +118,15 @@ class MobilerunKeyboardIME : InputMethodService() {
         return try {
             val ic = currentInputConnection
             if (ic != null) {
-                ic.sendKeyEvent(KeyEvent(KeyEvent.ACTION_DOWN, keyCode))
-                ic.sendKeyEvent(KeyEvent(KeyEvent.ACTION_UP, keyCode))
-                Log.d(TAG, "Direct key event sent: $keyCode")
-                true
+                val downAccepted = ic.sendKeyEvent(KeyEvent(KeyEvent.ACTION_DOWN, keyCode))
+                val upAccepted = ic.sendKeyEvent(KeyEvent(KeyEvent.ACTION_UP, keyCode))
+                val accepted = downAccepted && upAccepted
```
[the remark this answers is not in the document]
Comment (RasulOs) [reply app/src/main/java/com/mobilerun/portal/input/MobilerunKeyboardIME.kt]: Fixed in c8d1d3d by reverting the direct key-event return-value change from this PR. Direct key handling is back to its pre-PR behavior, so a partially accepted DOWN/UP pair cannot newly return false and trigger accessibility replay.

File: app/src/main/java/com/mobilerun/portal/input/InputConnectionTextEditor.kt [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    private fun matches(actual: String, expected: String, clear: Boolean): Boolean {
+        val normalizedActual = normalize(actual)
+        val normalizedExpected = normalize(expected)
+        return if (clear) {
+            normalizedActual == normalizedExpected
+        } else {
+            normalizedActual.contains(normalizedExpected)
```
[the remark this answers is not in the document]
Comment (RasulOs) [reply app/src/main/java/com/mobilerun/portal/input/InputConnectionTextEditor.kt]: Fixed in c8d1d3d. Append verification now builds the exact expected text by replacing the original selected range and compares the complete normalized result. A regression test confirms an unchanged pre-existing substring is not treated as success.

**Comment by RasulOs** [comment]
@codex review

File: app/src/main/java/com/mobilerun/portal/input/InputConnectionTextEditor.kt [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+        return snapshot.text.replaceRange(selectionStart, selectionEnd, text)
+    }
+
+    private fun canVerify(
+        before: Snapshot,
+        after: Snapshot,
+        clear: Boolean,
+        sameInputSession: Boolean,
+    ): Boolean {
+        if (clear) return after.coversEntireField
```
[the remark this answers is not in the document]
Comment (RasulOs) [reply app/src/main/java/com/mobilerun/portal/input/InputConnectionTextEditor.kt:276]: Fixed in a9e4442. The editor captures the original input-session generation and checks it before and after every connection operation and retry. A new field or finished input returns InputSessionChanged, and ApiHandler explicitly skips accessibility fallback for that result. IME restarts on the same editor remain retryable. Regression tests assert the original field is committed once and no retry/fallback reaches the next session.

File: app/src/main/java/com/mobilerun/portal/input/InputConnectionTextEditor.kt [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+            get() = startOffset == 0 && isCompleteReport
+    }
+
+    fun inputText(text: String, clear: Boolean): TextInputResult {
+        var clearAcceptedWithoutVerification = false
+
+        repeat(MAX_ATTEMPTS) { attempt ->
+            val connection = connectionProvider()
+            val before = connection?.let(::readSnapshot)
+            if (connection == null || before == null || (clear && !before.coversEntireField)) {
```
[the remark this answers is not in the document]
Comment (RasulOs) [reply app/src/main/java/com/mobilerun/portal/input/InputConnectionTextEditor.kt]: Fixed in a9e4442. For clear=false with no readable ExtractedText, the editor now calls commitText once at the connection's current selection and returns AcceptedUnverified when accepted. It does not call setSelection, retry the accepted append, or invoke accessibility fallback. Added direct-commit and API no-fallback regression coverage.

**Comment by RasulOs** [comment]
@codex review

File: app/src/main/java/com/mobilerun/portal/input/MobilerunKeyboardIME.kt [hunk 7]
```diff
@@ -94,29 +111,7 @@ class MobilerunKeyboardIME : InputMethodService() {
...
-                }
-            } else {
-                Log.w(TAG, "No input connection available for direct clear")
-                false
-            }
-        } catch (e: Exception) {
-            Log.e(TAG, "Error in direct text clear", e)
-            false
-        }
+        return inputTextResult("", clear = true) == TextInputResult.Verified
```
[the remark this answers is not in the document]
Comment (RasulOs) [reply app/src/main/java/com/mobilerun/portal/input/MobilerunKeyboardIME.kt]: Fixed in `eb3ebd1`. `clearTextResult()` now carries `Verified`, `Rejected`, `AcceptedUnverified`, or `InputSessionChanged` into `keyboardClear()`. Accessibility fallback runs only after a definite rejection; accepted or session-changed clears return an error without replay. Added API regression tests for all four outcomes and verified the dedicated clear endpoint on Android 16.

**Comment by RasulOs** [comment]
@codex review

File: app/src/main/java/com/mobilerun/portal/input/InputConnectionTextEditor.kt [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+            val expected = expectedText(before, text, clear)
+            if (!connection.setSelection(selection.first, selection.second)) {
+                if (!isSameInputSession(inputSessionGeneration)) {
+                    return TextInputResult.InputSessionChanged
+                }
+                if (attempt < MAX_ATTEMPTS - 1) {
+                    sleep(retryDelayMs)
+                    return@repeat
+                }
+                return TextInputResult.Rejected
```
[the remark this answers is not in the document]
Comment (RasulOs) [reply app/src/main/java/com/mobilerun/portal/input/InputConnectionTextEditor.kt:130]: Fixed in `b29890c`. Final retry failures from `finishComposingText`, `setSelection`, or `commitText` now preserve `AcceptedUnverified` when an earlier clear was accepted, so neither input endpoint nor clear endpoint can replay through accessibility. Added separate regression tests for rejected retry selection and rejected retry commit.

**Comment by RasulOs** [comment]
@codex review

File: app/src/main/java/com/mobilerun/portal/input/InputConnectionTextEditor.kt [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+            if (clear) {
+                clearAcceptedWithoutVerification = true
+            }
+
+            sleep(retryDelayMs)
+            if (!isSameInputSession(inputSessionGeneration)) {
+                return TextInputResult.InputSessionChanged
+            }
+            val afterConnection = connectionProvider()
+            val after = afterConnection?.let(::readSnapshot)
```
[the remark this answers is not in the document]
Comment (RasulOs) [reply app/src/main/java/com/mobilerun/portal/input/InputConnectionTextEditor.kt]: Fixed in `ad8f635` and mirrored internally in `eb2a328`. Connection acquisition and extracted-text reads are now contained inside the editor; after any accepted commit, null, stale, mismatched, or throwing readback returns `AcceptedUnverified` without retry or accessibility fallback. Added throwing post-state tests for both append and clear, each asserting exactly one commit.

**Comment by RasulOs** [comment]
@codex review

**Comment by RasulOs** [comment]
@codex review

File: app/src/main/java/com/mobilerun/portal/input/MobilerunKeyboardIME.kt [hunk 10]
```diff
@@ -220,9 +222,17 @@ class MobilerunKeyboardIME : InputMethodService() {
 
     override fun onStartInput(attribute: android.view.inputmethod.EditorInfo?, restarting: Boolean) {
         super.onStartInput(attribute, restarting)
+        if (!restarting) {
+            inputGeneration++
+        }
```
[the remark this answers is not in the document]
Comment (RasulOs) [reply app/src/main/java/com/mobilerun/portal/input/MobilerunKeyboardIME.kt]: Fixed in `4f8266f` and mirrored internally in `7f7d74f`. Every `onStartInput` callback now advances the generation, including `restarting=true`, before the framework hook. An in-flight operation therefore cannot treat a replacement connection from `restartInput` as its original session. Added callback-boundary regression tests in both repositories.

**Comment by RasulOs** [comment]
@codex review

File: app/src/main/java/com/mobilerun/portal/input/InputConnectionTextEditor.kt [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+        val startOffset: Int,
+        val partialStartOffset: Int,
+        val selectionStart: Int,
+        val selectionEnd: Int,
+    ) {
+        val isCompleteReport: Boolean
+            get() = partialStartOffset == -1
+
+        val coversEntireField: Boolean
+            get() = startOffset == 0 && isCompleteReport
```
[the remark this answers is not in the document]
Comment (RasulOs) [reply app/src/main/java/com/mobilerun/portal/input/InputConnectionTextEditor.kt]: Fixed in `b1bd641` and mirrored internally in `2188449`. For `clear=true`, a structurally complete extraction must now also prove it reaches the field end: `getTextAfterCursor` must exactly equal the suffix represented after the snapshot selection. Null, mismatch, or extra text rejects before `setSelection`/`commitText`, allowing the safe fallback. Added a truncated-suffix regression test and reran the Android 16 replacement/clear smoke.

**Comment by RasulOs** [comment]
@codex review

File: app/src/main/java/com/mobilerun/portal/input/InputConnectionTextEditor.kt [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+        text: String,
+        selectionStart: Int,
+        selectionEnd: Int,
+    ): Boolean {
+        val cursor = maxOf(selectionStart, selectionEnd)
+        val representedSuffix = text.substring(cursor)
+        val actualSuffix = connection.getTextAfterCursor(representedSuffix.length + 1, 0)
+            ?.toString()
+            ?: return false
+        return actualSuffix == representedSuffix
```
[the remark this answers is not in the document]
Comment (RasulOs) [reply app/src/main/java/com/mobilerun/portal/input/InputConnectionTextEditor.kt:254]: No code change: the short suffix read is an additional consistency check, not the sole field-boundary contract. We already require `ExtractedText.startOffset == 0` and `partialStartOffset == -1`; Android documents `-1` as “the content is the full text,” while `startOffset` anchors it in the overall text. `InputConnection.getTextAfterCursor` explicitly permits a return shorter than requested, so rejecting every short response would reject every normal cursor-at-EOF field (`n=1` correctly returns an empty string). If a custom editor both truncates/redacts surrounding text and falsely labels a truncated extraction as full text, no `InputConnection` read API can positively distinguish that contract violation. The current guard still rejects null, mismatched, or exposed extra suffix text without breaking compliant fields.

**Comment by RasulOs** [comment]
Physical Android 16 validation completed against head `b1bd641fb0fec6cd4af1a9eb30526b040e9f11f5`.

Device: Samsung SM-S918B, Android 16/API 36
WhatsApp: `com.whatsapp` 2.26.27.85
Debug APK SHA-256: `a79fd1dcbb86d4004c49b251be00a39abef78b5e8490813645a15ab783e53a12`

| Scenario | Result |
| --- | --- |
| Settings atomic replacement | PASS: `OLD7319` -> exactly `MRTEST7319`; response: `input done via IME (clear=true)` |
| Settings append | PASS: `MRTEST` + `7319` -> exactly `MRTEST7319`; response: `input done via IME (clear=false)` |
| Settings clear | PASS: field returned to hint-only `Search`; response: `Text cleared via IME` |
| IME restart/reacquisition | PASS: switched Portal -> Samsung -> Portal, then replaced with exactly `MRRESTART7319` |
| WhatsApp global search | PASS: 3 consecutive replacements with `9911656022`; each response reported IME success; raw text was `U+200B` + requested digits and normalized exactly |
| WhatsApp contact search | PASS: 2 consecutive replacements with exactly `9911656022`; each response reported IME success |
| WhatsApp Message yourself composer | PASS: exactly `MRTEST7319`, then cleared via IME; no message sent |

No accessibility fallback, false success, false negative, stale suffix, or duplicate append was observed. All test text was cleared. The phone was restored to Samsung Keyboard with accessibility disabled; both debug Portal apps remain installed but inactive.

**Comment by RasulOs** [comment]
Android 16 emulator edge-case validation completed against source head `b1bd641fb0fec6cd4af1a9eb30526b040e9f11f5` on `emulator-5554` (API 36).

Result: **140/140 runtime cases passed**.

Covered:
- shorter, longer, and empty atomic replacements;
- empty and normal appends;
- cursor insertion (`ABCDE` -> `ABCXDE`);
- selected-range replacement (`ABCDE`, select `CD`, insert `XY` -> `ABXYE`);
- spaces, punctuation, emoji, combining characters, Arabic, requested U+200B/U+FEFF, multiline text, and an exact 4096-character value;
- 50 alternating replacements, 50 sequential appends, and 20 repeated clears;
- no-focused-editor rejection;
- inactive-IME accessibility fallback reporting;
- numeric Google Messages recipient input and unsent multiline composer input;
- focus-change race: 1 induced session change, 0 wrong-field writes;
- IME-restart sweep at 0–90 ms: 8 session-change errors with fallback skipped, 2 completed successes, exact `BASEX` in all 10 outcomes, and 0 duplicate appends.

Repository checks:
- Focused input/API unit tests and `assembleDebug` pass.
- The complete `testDebugUnitTest` run still has only the pre-existing unrelated `SwitchTintResourceTest.trackTintSeparatesCheckedAndUncheckedStates` tint-resource failure.

No public Portal defect was found and no code change was needed. Test text was cleared, no message was sent, and no draft remains.
